src/loaders/JsonImageInfoLoader.ts

```diff
@@ -187,57 +198,90 @@ class JsonImageInfoLoader implements IVolumeLoader {
       const url = imageArray[i].name;
       const batch = imageArray[i].channels;
 
-      // using Image is just a trick to download the bits as a png.
-      // the Image will never be used again.
-      const img: HTMLImageElement = new Image();
-      img.onerror = function () {
-        console.log("ERROR LOADING " + url);
+      const regionPx = convertSubregionToPixels(volume.loadSpec.subregion, volume.imageInfo.subregionSize);
+      const cacheQueryDims = {
+        region: regionPx,
+        time: volume.loadSpec.time,
+        scale: 0,
       };
-      img.onload = (function (thisbatch) {
-        return function (event: Event) {
-          //console.log("GOT ch " + me.src);
-          // extract pixels by drawing to canvas
-          const canvas = document.createElement("canvas");
-          // nice thing about this is i could downsample here
-          const w = Math.floor((event?.target as HTMLImageElement).naturalWidth);
-          const h = Math.floor((event?.target as HTMLImageElement).naturalHeight);
-          canvas.setAttribute("width", "" + w);
-          canvas.setAttribute("height", "" + h);
-          const ctx = canvas.getContext("2d");
-          if (!ctx) {
-            console.log("Error creating canvas 2d context for " + url);
-            return;
-          }
-          ctx.globalCompositeOperation = "copy";
-          ctx.globalAlpha = 1.0;
-          ctx.drawImage(event?.target as CanvasImageSource, 0, 0, w, h);
-          // getImageData returns rgba.
-          // optimize: collapse rgba to single channel arrays
-          const iData = ctx.getImageData(0, 0, w, h);
-
-          const channelsBits: Uint8Array[] = [];
-          // allocate channels in batch
-          for (let ch = 0; ch < Math.min(thisbatch.length, 4); ++ch) {
-            channelsBits.push(new Uint8Array(w * h));
-          }
-          // extract the data
-          for (let j = 0; j < Math.min(thisbatch.length, 4); ++j) {
-            for (let px = 0; px < w * h; px++) {
-              channelsBits[j][px] = iData.data[px * 4 + j];
+      // if any in batch is cached then they all are...?
+      let cacheHit = false;
+      for (let j = 0; j < Math.min(batch.length, 4); ++j) {
```

Comment:
If the first item is a hit, and the second is a miss, cacheHit is true, and no further items are checked?  This boolean flag seems odd inside the loop.  If you only want to see if at least one is a hit, break on if (cacheResult), not else.

Reply:
In this particular loader, the data is requested and arrives in batches of up to 4.  The reason why this code works is because it's order-independent because of the batched nature of the data.  Either all items in the batch will be cached or they won't.  So that assumption is built-in here.
Any cache miss means we need to request the whole batch.  I could keep looping, but it's going to fetch a whole batch of data anyway.  Maybe would be easier to read if I didn't break out of the loop.
(I could also rewrite this to check cache for batch[0] only, and then loop over batches based on that result.)

Reply:
We still need the loop in order to fetch the cached items, but yes.

Reply:
Ah.  I was focused on loop control, and missed that part.  Thanks for explaining!